### PR TITLE
Custom feature extractor

### PR DESCRIPTION
**Pull request by sgugger**
# What does this PR do?

This PR adds support for custom feature extractor in the Hub, with the same APIs as for models, configurations and tokenizers.

### CODE REVIEW

File: src/transformers/models/auto/feature_extraction_auto.py [hunk 1]
```diff
@@ -66,6 +71,96 @@ def feature_extractor_class_from_name(class_name: str):
...
+    pretrained_model_name_or_path: Union[str, os.PathLike],
+    cache_dir: Optional[Union[str, os.PathLike]] = None,
+    force_download: bool = False,
+    resume_download: bool = False,
+    proxies: Optional[Dict[str, str]] = None,
+    use_auth_token: Optional[Union[bool, str]] = None,
+    revision: Optional[str] = None,
+    local_files_only: bool = False,
+    **kwargs,
+):
```
Comment (LysandreJik) [inline src/transformers/models/auto/feature_extraction_auto.py:84]: Nice! Does this also add support for the feature extractors with a custom version set in the filename?

File: src/transformers/models/auto/feature_extraction_auto.py [hunk 2]
```diff
@@ -151,35 +250,54 @@ def from_pretrained(cls, pretrained_model_name_or_path, **kwargs):
...
 
-        if "feature_extractor_type" in config_dict:
-            feature_extractor_class = feature_extractor_class_from_name(config_dict["feature_extractor_type"])
             return feature_extractor_class.from_dict(config_dict, **kwargs)
-        elif model_type is not None:
-            return FEATURE_EXTRACTOR_MAPPING[type(config)].from_dict(config_dict, **kwargs)
+        # Last try: we use the FEATURE_EXTRACTOR_MAPPING.
+        elif type(config) in FEATURE_EXTRACTOR_MAPPING:
+            feature_extractor_class = FEATURE_EXTRACTOR_MAPPING[type(config)]
+            return feature_extractor_class.from_dict(config_dict, **kwargs)
```
Comment (LysandreJik) [inline src/transformers/models/auto/feature_extraction_auto.py:298]: Welcome change!

**Review by LysandreJik — APPROVED** [review]
Great work, looks good to me! Thanks for cleaning up this file!

File: src/transformers/models/auto/feature_extraction_auto.py (hunk 1 above)
Comment (sgugger) [reply src/transformers/models/auto/feature_extraction_auto.py:84]: No, this is something different.

**Pull request merged**